src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java

```diff
@@ -197,32 +231,49 @@ private void setupVCFWriter(VCFHeader inputVCFHeader, SampleList samples) {
     }
 
     @Override
-    public void apply(VariantContext variant, ReadsContext reads, ReferenceContext ref, FeatureContext features) {
+    public void apply(final Locatable loc, List<VariantContext> variants, ReadsContext reads, ReferenceContext ref, FeatureContext features) {
+
+        final List<VariantContext> variantsToProcess = getVariantSubsetToProcess(loc, variants);
+
         ref.setWindow(10, 10); //TODO this matches the gatk3 behavior but may be unnecessary
-        final VariantContext mergedVC = merger.merge(Collections.singletonList(variant), variant, includeNonVariants ? ref.getBase() : null, true, false);
+        final VariantContext mergedVC = merger.merge(variantsToProcess, loc, includeNonVariants ? ref.getBase() : null, !includeNonVariants, false);
         final VariantContext regenotypedVC = regenotypeVC(mergedVC, ref, features, includeNonVariants);
         if (regenotypedVC != null) {
             final SimpleInterval variantStart = new SimpleInterval(regenotypedVC.getContig(), regenotypedVC.getStart(), regenotypedVC.getStart());
-            if (!onlyOutputCallsStartingInIntervals || intervals.stream().anyMatch(interval -> interval.contains    (variantStart))) {
+            if (!GATKVariantContextUtils.isSpanningDeletionOnly(regenotypedVC) &&
+                    (!onlyOutputCallsStartingInIntervals || intervals.stream().anyMatch(interval -> interval.contains (variantStart)))) {
                 vcfWriter.add(regenotypedVC);
             }
         }
     }
 
-
+    // If includeNonVariants is set, we're using group-by-locus traversal. To match GATK3 GenotypeGVCFs,
+    // see if there is a variant in the overlapping group that starts exactly at the locus start position, and if so
+    // prioritize and process only that variant. Otherwise process all of the overlapping variants.
+    private List<VariantContext> getVariantSubsetToProcess(final Locatable loc, List<VariantContext> preProcessedVariants) {
+        return includeNonVariants ?
+                preProcessedVariants.stream().anyMatch(vc -> vc.getStart() == loc.getStart()) ?
```

Comment:
Multiple nested ternary expressions with multiple stream expressions are hard to parse, and it's a bit wasteful to run the same stream twice in a row to check if something exists and then take the first one.  Could you rewrite this so it's a little clearer and can reuse the value if it finds it.  I might also add a check to be sure that there is exactly 1, because we probably have a problem if there are more than 1.

Reply:
Done.